### PR TITLE
This loads a linux payload.

### PR DESCRIPTION
**Pull request by rminnich**

### CODE REVIEW

**Comment by rminnich** [comment]
I realize this code is not good at all. but ... 

it booted.

File: payloads/src/payload/payload.rs [hunk 1]
```diff
@@ -1,6 +1,11 @@
 use core::intrinsics::{copy, transmute};
 use model::{Driver, EOF};
-
+use postcard::from_bytes;
+use serde::Deserialize;
+use wrappers::{Memory, SectionReader};
+use core::fmt::Write;
+use print;
+//use serde::{Serialize, Deserialize};
```
Comment (rjoleary) [inline payloads/src/payload/payload.rs]: Delete this line

File: payloads/src/payload/payload.rs [hunk 2]
```diff
@@ -83,9 +88,36 @@ pub enum stype {
...
+    CBFS_SEGMENT_PARAMS = 0x41524150,
+    CBFS_SEGMENT_ENTRY = 0x52544E45,
 }
 
+// I give up.
+impl From<u32> for stype {
+    fn from(s: u32) -> Self {
+        match s {
+            0x434F4445 => stype::PAYLOAD_SEGMENT_CODE,
+            0x44415441=> stype::PAYLOAD_SEGMENT_DATA,
```
Comment (rjoleary) [inline payloads/src/payload/payload.rs]: Don't forget to run rustfmt.

File: payloads/src/payload/payload.rs [hunk 3]
```diff
@@ -83,9 +88,36 @@ pub enum stype {
...
+    PAYLOAD_SEGMENT_DTB = 0x44544220,
+    PAYLOAD_SEGMENT_BAD = 0xFFFFFFFF,
+    CBFS_SEGMENT_CODE = 0x45444F43,
+    CBFS_SEGMENT_DATA = 0x41544144,
+    CBFS_SEGMENT_BSS = 0x20535342,
+    CBFS_SEGMENT_PARAMS = 0x41524150,
+    CBFS_SEGMENT_ENTRY = 0x52544E45,
 }
 
+// I give up.
```
Comment (rjoleary) [inline payloads/src/payload/payload.rs]: But there's so much to live for.

File: payloads/src/payload/payload.rs [hunk 4]
```diff
@@ -83,9 +88,36 @@ pub enum stype {
...
+impl From<u32> for stype {
+    fn from(s: u32) -> Self {
+        match s {
```
Comment (rjoleary) [inline payloads/src/payload/payload.rs]: Maybe do what they suggest here? https://enodev.fr/posts/rusticity-convert-an-integer-to-an-enum.html

File: src/arch/x86/x86_64/src/bootblock.S [hunk 5]
```diff
@@ -351,7 +351,7 @@ lme:
 _identity:
 	// This is a software breakpoint, for use until we're more
 	// sure about the code. It is to be removed. This is the
-	1:	 jmp 1b
+//	1:	 jmp 1b
```
Comment (rjoleary) [inline src/arch/x86/x86_64/src/bootblock.S]: Delete this line to make it look cleaner.

File: src/drivers/uart/src/null.rs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 // null is a simple Null driver.
 // TODO: move it out of uart.
 use model::*;
+use x86::io::{outb};
```
Comment (rjoleary) [inline src/drivers/uart/src/null.rs]:
```suggestion
use x86::io::outb;
```

File: src/drivers/uart/src/null.rs [hunk 7]
```diff
@@ -14,6 +15,9 @@ impl Driver for Null {
     }
 
     fn pwrite(&mut self, data: &[u8], _offset: usize) -> Result<usize> {
+        for (_i, &c) in data.iter().enumerate() {
+            unsafe {outb(0x3f8, c);}
```
Comment (rjoleary) [inline src/drivers/uart/src/null.rs]: Isn't the Null driver just not supposed to do anything?

File: src/mainboard/emulation/qemu-q35/Makefile.toml [hunk 8]
```diff
@@ -0,0 +1,45 @@
+[config]
```
Comment (rjoleary) [inline src/mainboard/emulation/qemu-q35/Makefile.toml]: Did the Makefile not work? :(

File: src/mainboard/emulation/qemu-q35/fixed-dtfs.dts [hunk 9]
```diff
@@ -57,14 +57,15 @@
                 compatible = "ore-rampayload";
                 half = <0>;
                 offset = <0x0>;
-                size = <0x600000>;
+                size = <0xc00000>;
+		file = "payloadA";
```
Comment (rjoleary) [inline src/mainboard/emulation/qemu-q35/fixed-dtfs.dts]: Indentation is :boom:

File: src/arch/x86/x86_64/src/ioport.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+// ioports reads and writes bytes to ioports, using the offset.
```
Comment (rjoleary) [inline src/arch/x86/x86_64/src/ioport.rs:1]: This module isn't included anywhere. You could likely just dump this code inside the lib.rs file.

**Comment by rminnich** [comment]
seg CBFSSeg { typ: 1381256773, comp: 0, off: 0, load: 4294967296, len: 0, memlen: 0 }
afterward seg CBFSSeg { typ: 1381256773, comp: 0, off: 0, load: 16777216, len: 0, memlen: 0 }
ENTRY 16777216
Running payload
on to 0x1000000Linux version 5.5.0-rc3+ (rminnich@xcpu) (gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)) #111 Wed Feb 19 18:38:53 UTC 2020
Command line: 
x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
x86/fpu: Supporting XSAVE feature 0x200: 'Protection Keys User registers'
x86/fpu: xstate_offset[9]: 2688, xstate_sizes[9]:    8
x86/fpu: Enabled xstate features 0x203, context size is 2696 bytes, using 'standard' format.
BIOS-provided physical RAM map:
BIOS-e801: [mem 0x0000000000000000-0x000000000009efff] usable
printk: console [earlyser0] enabled
NX (Execute Disable) protection: active
tsc: Fast TSC calibration using PIT
tsc: Detected 2207.410 MHz processor
.text .data .bss are not marked as E820_TYPE_RAM!
last_pfn = 0x1a16 max_arch_pfn = 0x400000000
x86/PAT: Configuration [0-7]: WB  WT  UC- UC  WB  WT  UC- UC  
Using GB pages for direct mapping
Kernel panic - not syncing: alloc_low_pages: can not alloc memory

File: src/drivers/uart/src/i8250.rs [hunk 11]
```diff
@@ -0,0 +1,138 @@
+// http://www.ti.com/lit/ds/symlink/pc16550d.pdf
```
Comment (fotonick) [inline src/drivers/uart/src/i8250.rs:1]: This link is broken. It'd be really nice to have an up-to-date reference document that you're coding against.

File: src/drivers/uart/src/i8250.rs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+//use core::ops;
```
Comment (fotonick) [inline src/drivers/uart/src/i8250.rs:2]: In general, this file has way more commented-out code than is usually justified. Most code standards don't want to see any. Is all of it valid and interesting?

Comment (rminnich) [reply src/drivers/uart/src/i8250.rs:2]: for an early stage project like this, I'm ok with it. It's a sort of TODO pointer. Later on, standards will have to be much stricter.

File: src/drivers/uart/src/i8250.rs (hunk 11 above)
Comment (rminnich) [reply src/drivers/uart/src/i8250.rs:1]: broken? Damn. It's from coreboot. I argue with them all the time about not committing links, and here's why. I'll remove it.

File: payloads/src/payload/payload.rs [hunk 13]
```diff
@@ -116,6 +171,111 @@ pub struct Segment<'a> {
...
+                }
+                _ => {
+                    write!(w, "Seg is unchanged: {:?}\n", seg).unwrap();
+                }
+            }
+
+            let mut load = seg.load as usize;
+
+            // Copy from driver into segment.
+            let mut buf = [0u8; 512];
```
[the remark this answers is not in the document]
Comment (rminnich) [reply payloads/src/payload/payload.rs:230]: It's a balance between stack size and performance. 4096 is a bit high (a full page) but 512 has proven to be sort of ok.

File: payloads/src/payload/payload.rs [hunk 14]
```diff
@@ -116,6 +171,111 @@ pub struct Segment<'a> {
...
+                stype::CBFS_SEGMENT_ENTRY
+                | stype::CBFS_SEGMENT_CODE
+                | stype::CBFS_SEGMENT_DATA
+                | stype::CBFS_SEGMENT_BSS
+                | stype::CBFS_SEGMENT_PARAMS => {
+                    write!(w, "seg {:?}\n", seg).unwrap();
+                    seg.off = seg.off.to_be();
+                    seg.load = seg.load.to_be();
+                    seg.len = seg.len.to_be();
+                    seg.memlen = seg.memlen.to_be();
```
[the remark this answers is not in the document]
Comment (rminnich) [reply payloads/src/payload/payload.rs:216]: this code is horrible, It's another "this works, now make it decent".

I've got to remember why it goes to be. But it really needs to be cleaned up.

File: src/drivers/uart/src/i8250.rs (hunk 11 above)
Comment (rminnich) [reply src/drivers/uart/src/i8250.rs]: removed.

**Comment by rminnich** [comment]
NO need to approve this. comments and suggestions to fix it are best at this point. Consider this a storage mechanism.

**Pull request merged**